Sources/SwiftProtobuf/JSONEncoder.swift

```diff
@@ -53,68 +129,116 @@ public struct JSONEncoder {
         } else {
             // TODO: Be smarter here about choosing significant digits
             // See: protoc source has C++ code for this with interesting ideas
-            let s: String
-            if value < Double(Int64.max) && value > Double(Int64.min) && value == Double(Int64(value)) {
-                s = String(Int64(value))
-            } else {
-                s = String(value)
-            }
-            if quote {
-                append(text: "\"" + s + "\"")
+            if let v = Int64(exactly: value) {
+                appendInt(value: v)
             } else {
+                let s = String(value)
                 append(text: s)
             }
         }
     }
-    mutating func putInt64(value: Int64, quote: Bool) {
-        // Always quote integers with abs value > 2^53
-        if quote || value > 0x1FFFFFFFFFFFFF || value < -0x1FFFFFFFFFFFFF {
-            append(text: "\"" + String(value) + "\"")
+    private mutating func appendUInt(value: UInt64) {
```

Comment:
> ... I get kind of an "assembly language in Swift syntax" feeling ...

Agreed.  My primary goals for this PR are to:
1. get the API to a place where very good performance is possible.   The easiest way to prove that is to actually demonstrate such performance.
2. isolate the ugly bits.

If we can achieve the first goal, it means that we should be able to commit to the generated code API for a long time.  If we can achieve the second, we'll be able to iterate on better ways to express these things.  I also expect that such games will be less necessary over time as the Swift library implementation continues to improve.

Right now, this PR still has a flaw in exposing the `JSONEncoder` and the associated `FieldType` bits.  I think I see a way to fix that but it might take me a few days to work through the details...